addons/website_sale_stock/controllers/main.py

```diff
@@ -26,3 +27,28 @@ def shop_payment_transaction(self, *args, **kwargs):
         if values:
             raise ValidationError(' '.join(values))
         return super().shop_payment_transaction(*args, **kwargs)
+
+
+class WebsiteSale(website_sale_controller.WebsiteSale):
+    @http.route(['/shop/add/stock_notification'], type="json", auth="public", website=True)
+    def add_stock_email_notification(self, email, product_id):
+        if not email_re.match(email):
+            raise BadRequest(_("Invalid Email"))
+
+        product = request.env['product.product'].browse(int(product_id))
+        partners = request.env['res.partner'].sudo()._mail_find_partner_from_emails([email], force_create=True)
+        partner = partners[0]
+
+        if not product._has_stock_notification(partner):
+            product.sudo().stock_notification_partner_ids += partner
+
+        if request.website.is_public_user():
+            request.session.setdefault('stock_notifications', []).append(product_id)
+            request.session['stock_notification_email'] = email
+            request.session.touch()
```

Comment:
We should avoid at all cost writing to the session, because of that we can take advantage of session for checking if the value has changed. I would also use a set to store the `stock_notifications` instead.
That way if for some reason none of those values end up changing it will not have to save anything
Also note that I am not convinced `stock_notifications` is the right name for this as it is a list of product ids.
(I haven't checked if the session supports sets but I don't see why it couldn't.)
```suggestion
            request.session['stock_notifications'] = request.session.get('stock_notifications', set()) | {product_id}
            request.session['stock_notification_email'] = email
```